### PR TITLE
feat: chordpro & chords over words compatibility (WIP)

### PR DESCRIPTION
**Pull request by isaiahdahl**
- [x] basic chords over words formatter
- [x] basic lyric space trimming on the chords over words parser
- [x]  handle the scenario where there are two chords only lines back to back
- [x]  trimming the spacing correctly when the previous chord pushes it into the void between two chords.
- [x]  handle chords where only the base is specified ie: [C/E][/D][/Db]
- [ ] handle lyrics that start with a valid chord (for example `I searched the stars`)

The goal is to take a very complicated song, go from chordpro, to chords over words, and back to chordpro and have it be exactly the same.

### CODE REVIEW

**Comment by isaiahdahl** [comment]
@martijnversluis Ok I'm ready for you to take an attempt at the tests I have failing.

I've added one of the more complicated songs I can find and written some test around it. I made some adjustments to get some of the new tests I added to pass but there's a few I'm stuck on.

I also added the chords over words formatter copying it from the text fomatter. I did a very very rudimentary change so that it renders more than just the title for metadata, it's definitely a WIP file and should be written as if it's designed to be the chords over words formatter, but it gets the studio prototype working a bit more.

We don't have syntax highlighting on the chords over words but we're getting closer to the experience of switching between chordpro and chords over words! Here's a little something I was able to get working locally. 

CleanShot 2022-10-24 at 14 07 55

This is an overview of the ones I couldn't get passing:
---
- [ ] handle the scenario where there are two chords only lines back to back

If you take a Rhythm section like this
```
Interlude 4 *18
Dm9 / C/E / | F9/A Bbma9 / A7(#9#5) |
Dm9 / C/E /D - /Db - /C | F13 Bbma9 / A7(#9#5) |
```
It thinks the second line is lyrics for the first line.

---  

- [ ] trimming the spacing correctly when the previous chord pushes it into the void between two chords.

This chordpro:
```
We’ll be [Dm7/A]singing for - [G13]ever and [G13(#5)]ever,[Gm7/C] a - [F]men
```
generates these chords over words:
```
         Dm7/A         G13      G13(#5) Gm7/C     F
We’ll be singing for - ever and ever,         a - men
```
So the chords over words generate back to the same chordpro. 

The weird problem to figure out is how to not put a space after `ever,` 
The `G13(#5)` pushes out the `Gm7/C` into the void between `ever,` and `amen`

It's displayed correctly in the chords over words, but somehow we gotta figure out how to turn it back into chordpro that makes the `Gm7/C` a trailing chord to the `ever,` lyric without a space after it, while not messing up other scenarios where it should be trimmed to a single trailing space.

---  

- [ ] handle chords where only the base is specified ie: `[C/E][/D][/Db]` 

I've discovered a fairly common scenario in our songs where only the bass note is specified. It makes for a cleaner chart when the bass note is walking around. We should be able to support parsing a chord where only the bass is specified.

**Comment by isaiahdahl** [comment]
Hey @martijnversluis  Just want to touch base on this one. 

Have you had a look at my comments above?

From the three tasks I highlighted in the comment above would you prefer to break those out into separate issues/PRs?

**Comment by martijnversluis** [comment]
Hey @isaiahdahl. I totally missed this PR. I'm gonna put it on the RFC tracker board (which is currently what I use to determine what to work on), and take a look at it 👍

Will get back to you 😉

File: src/formatter/templates/html_div_formatter.ts [hunk 1]
```diff
@@ -44,7 +44,7 @@ export default (
                 ${ when(isChordLyricsPair(item), () => `
                   <div class="column">
                     ${ keep([renderChord(item.chords, line, song, key)], ([renderedChord]) => `
-                      <div class="chord"${ renderedChord ? fontStyleTag(line.chordFont) : '' }>${ renderedChord }</div>
+                      <div class="chord"${ renderedChord ? fontStyleTag(line.chordFont) : '' }>${ renderedChord } </div>
```
Comment (martijnversluis) [inline src/formatter/templates/html_div_formatter.ts:47]: Does this change also belong in this PR, or is it accidental?

Comment (isaiahdahl) [reply src/formatter/templates/html_div_formatter.ts:47]: Semi-accidental in this PR. It was a fix to create better spacing between trailing chords as well as add some space between chords in a rhythm only section so it doesn't render like this:

```
[Eb5][/][/][/][|][/][/][/][/]
```
CleanShot 2022-12-12 at 09 20 22

I had started this PR a while ago: https://github.com/bettermusic/ChordSheetJS/pull/259 which was for the same thing, this was just a quick approach to make the studio renderer feel more realistic.

**Comment by martijnversluis** [comment]
Linking this issue as I'm currently experiencing the same issue during the refactor: https://github.com/martijnversluis/ChordSheetJS/issues/821

**Comment by martijnversluis** [comment]
I'm gonna merge this because it's open for too long contains other bugfixes that can already be released. There is one pending TODO on the lyric words that are valid chords, I will address that in a separate PR.

**Pull request merged**